examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/Logging.cs

```diff
@@ -0,0 +1,127 @@
+           1:IP   2:client   3:user   4:date time   5:method   
+           6:req   7:proto   8:respcode   9:size */
+        static readonly string s_apacheRx =
+                "^(\\S+) (\\S+) (\\S+) \\[([\\w:/]+\\s[+\\-]\\d{4})\\] \"(\\S+) (\\S+) (\\S+)\" (\\d{3}) (\\d+)";
+
+        public void Run(string[] args)
+        {
+            if (args.Length != 1)
+            {
+                Console.Error.WriteLine(
+                    "Usage: Logging <path to Apache User Logs>");
+                Environment.Exit(1);
+            }
+
+            SparkSession spark = SparkSession
+                .Builder()
+                .AppName("Apache User Log Processing")
+                .GetOrCreate();
+
+            // Read input log file and display it
+            var df = spark.Read().Text(args[0]);
+            df.Show();
+
+            // Step 1: UDF to determine if each line is a valid log entry
+            // Remove any invalid entries before further filtering
+            spark.Udf().Register<string, bool>(
+                "GeneralReg",
+                log => Regex.IsMatch(log, s_apacheRx));
+
+            df.CreateOrReplaceTempView("Logs");
+
+            // Apply the UDF to get valid log entries
+            DataFrame generalDf = spark.Sql(
+                "SELECT logs.value, GeneralReg(logs.value) FROM Logs");
+
+            // Only keep log entries that matched the reg ex
+            generalDf = generalDf.Filter(generalDf["GeneralReg(value)"]);
+            generalDf.Show();
+
+            // View the resulting schema
+            // Notice we created a new column "GeneralReg(value)"
+            generalDf.PrintSchema();
+
+            // Step 2: Choose valid log entries that start with 10
+            spark.Udf().Register<string, bool>(
+                "IPReg",
+                log => Regex.IsMatch(log, "^(?=10)"));
+
+            generalDf.CreateOrReplaceTempView("IPLogs");
+
+            // Apply UDF to get valid log entries starting with 10
+            // Use SQL "WHERE" rather than doing ipDf.Filter(),
+            // which avoids creating an extra column "IPReg(value)"
+            DataFrame ipDf = spark.Sql(
+                "SELECT iplogs.value FROM IPLogs WHERE IPReg(iplogs.value)");
+            ipDf.Show();
+
+            // Step 3: Choose valid log entries that start 
+            // with 10 and deal with spam
+            spark.Udf().Register<string, bool>(
+                "SpamRegEx",
+                log => Regex.IsMatch(log, "\\b(?=spam)\\b"));
+
+            ipDf.CreateOrReplaceTempView("SpamLogs");
+
+            // Apply UDF to get valid, start with 10, spam entries
+            DataFrame spamDF = spark.Sql(
+                "SELECT spamlogs.value FROM SpamLogs WHERE SpamRegEx(spamlogs.value)");
+
+            // Explore the columns in the data we have filtered
+            // Let's try getting the number of GET requests
+            IEnumerable<Row> rows = spamDF.Collect();
```

Comment:
nit.  Can we change this to a linq query ?  Something like
```csharp
int numGetRequests = spamDF.Collect().Filter(r => ContainsGet(r.GetAs<string>("value"))).Count();
```

Reply:
Looking at converting it to LINQ, and issue is that I end up with an `IEnumerable` of rows that aren't compatible with LINQ:

int numGetRequests = spamDF
                .Collect()
                .Where(r => ContainsGet(r.GetAs<string>("value"))).Count();

produces error:

`'IEnumerable<Row>' does not contain a definition for 'Where' and no accessible extension method 'Where' accepting a first argument of type 'IEnumerable<Row>' could be found`

I'd have to cast to a more generic IEnumerable, so may not be that much cleaner with LINQ instead.

Reply:
Wasn't `using` LINQ at the top, so once I added that and made a couple of small adjustments, got it to work 👍